metal/resource_metal_device.go

```diff
@@ -649,9 +705,55 @@ func resourceMetalDeviceUpdate(d *schema.ResourceData, meta interface{}) error {
 		}
 
 	}
+
+	if d.HasChange("operating_system") || d.HasChange("user_data") || d.HasChange("custom_data") {
+		reinstallOptions, err := getReinstallOptions(d)
+
+		if err != nil {
+			return friendlyError(err)
+		}
+
+		if _, err := client.Devices.Reinstall(d.Id(), &reinstallOptions); err != nil {
```

Comment:
Terraform providers should not return a device until it is in a usable state.
We will have to do a wait after calling Reinstall (as we do after Create).

We may have to teach the wait loop that `reinstalling` is an expected transitional state.

Reply:
We may need a different Wait function if the existing one is not configurable (if we can't give it the desired state, and errors states).

We may also want to check the Update timeout limits to see if this change will cause Terraform to bump into those limits.  I can't think of any other long-running Update processes today (ignoring ForceNew).

Reply:
Fixed in new commit.